src/katgpucbf/xbgpu/engine.py

```diff
@@ -117,6 +121,426 @@ def reset(self, timestamp: int = 0) -> None:
+                "For the latest accumulation, was data present from all F-Engines.",
+                default=False,
+                initial_status=aiokatcp.Sensor.Status.ERROR,
+                status_func=lambda value: aiokatcp.Sensor.Status.NOMINAL if value else aiokatcp.Sensor.Status.ERROR,
+            )
+        )
+        sensors.add(
+            aiokatcp.Sensor(
+                int,
+                f"{self.output.name}.xeng-clip-cnt",
+                "Number of visibilities that saturated",
+                default=0,
+                initial_status=aiokatcp.Sensor.Status.NOMINAL,
+            )
+        )
+
+    async def _flush_accumulation(self, tx_item: TxQueueItem, next_accum: int) -> TxQueueItem:
+        """Emit the current `tx_item` and prepare a new one."""
+        if tx_item.batches == 0:
+            # We never actually started this accumulation. We can just
+            # update the timestamp and continue using it.
+            tx_item.timestamp = next_accum * self.timestamp_increment_per_accumulation
+            return tx_item
+
+        # present_ants only takes into account batches that have
+        # been seen. If some batches went missing entirely, the
+        # whole accumulation is bad.
+        if tx_item.batches != self.output.heap_accumulation_threshold:  # type: ignore
+            tx_item.present_ants.fill(False)
+
+        # Update the sync sensor (converting np.bool_ to Python bool)
+        self.engine.sensors[f"{self.output.name}.rx.synchronised"].value = bool(tx_item.present_ants.all())
+
+        self.correlation.reduce()
+        tx_item.add_marker(self._proc_command_queue)
+        await self._tx_item_queue.put(tx_item)
+
+        # Prepare for the next accumulation (which might not be
+        # contiguous with the previous one).
+        tx_item = await self._tx_free_item_queue.get()
+        await tx_item.async_wait_for_events()
+        tx_item.timestamp = next_accum * self.timestamp_increment_per_accumulation
+        self.correlation.bind(out_visibilities=tx_item.buffer_device, out_saturated=tx_item.saturated)
+        self.correlation.zero_visibilities()
+        return tx_item
+
+    async def gpu_proc_loop(self) -> None:  # noqa: D102
+        """Perform all GPU processing of received data in a continuous loop.
+
+        This function performs the following steps:
+        1. Retrieve an rx_item from the _rx_item_queue
+        2.1 Apply the correlation kernel to small subsets of the data
+            until all the data has been processed.
+        2.2 If sufficient correlations have occured, transfer the correlated
+            data to a tx_item, pass the tx_item to the _tx_item_queue and get a
+            new item from the _tx_free_item_queue.
+
+        The ratio of rx_items to tx_items is not one to one. There are expected
+        to be many more rx_items in for every tx_item out.
+        """
+        rx_item: RxQueueItem | None
```

Comment:
Trust me, I didn't enjoy adding it either 😅 Something in the pre-commit was complaining it didn't know what `rx_item` is below/in the helper function.

The removal results in "Cannot determine type of `rx_item`"
* Maybe I'll add a `type: ignore` below?

Reply:
While it's not pretty, having mypy know the types of things is definitely preferable to sprinkling on `type: ignore`.

Reply:
I'm surprised that mypy can't figure it out because it knows what the type of `_rx_item_queue` is, so it should know what kind of thing `get()` produces.

Would it not be possible to add the type hint where `rx_item` is below? like
```
rx_item: RxQueueItem | None = await self._rx_item_queue.get()
```

I didn't test that at all, this is just how I would have done it. What are the down-sides to what I've suggested compared with what you've got here?

(This is a really minor item. Ignore if you even have the slightest inkling of not wanting to bother.)